.travis.yml

```diff
@@ -1,25 +1,125 @@
+      - wget https://github.com/ocaml/opam/releases/download/$VERSION/$PKG
+      - sudo mv $PKG /usr/local/bin/opam
+      - sudo chmod a+x /usr/local/bin/opam
+
+      - opam init -y --bare --disable-sandboxing --disable-shell-hook
+      - CACHED=yes
+      - '[ -d _opam/bin ] || CACHED=no'
+      - '[ "$CACHED" == yes ] || rm -rf _opam'
+      - '[ "$CACHED" == yes ] || opam switch create . $COMPILER $REPOSITORIES --no-install'
+      - eval `opam env`
+      - opam --version
+      - ocaml -version
+
+      - opam pin add --no-action ounit 2.0.8 # 2.1.2 release is broken.
+      - CACHED=yes
+      - '[ -d _cache/_build ] || CACHED=no'
+      - '[ "$CACHED" == yes ] || opam install . --deps-only --yes'
+      - '[ "$CACHED" == yes ] || [ "$LIBEV" == no ] || opam install -y conf-libev'
+      - '[ "$CACHED" == no ] || cp -r _cache/_build .'
+
+      - '[ "$LIBEV" == no ] || LWT_DISCOVER_ARGUMENTS="--use-libev true"'
+      - '[ "$LIBEV" != no ] || LWT_DISCOVER_ARGUMENTS="--use-libev false"'
+      - export LWT_DISCOVER_ARGUMENTS
+
+    script:
+      - '[ "$COVERAGE" == yes ] || dune build'
+      - '[ "$COVERAGE" == yes ] || dune runtest --force'
+      - '[ "$COVERAGE" != yes ] || make coverage-only'
+      - '[ "$COVERAGE" != yes ] || bisect-ppx-report send-to Coveralls'
+
+      - '[ "$CACHED" == yes ] || mkdir -p _cache'
+      - '[ "$CACHED" == yes ] || cp -r _build _cache'
+
+      - '[ "$DOCS" != yes ] || bash src/util/travis-docs.sh'
+      - '[ "$PACKAGING" != yes ] || make install-for-packaging-test packaging-test uninstall-after-packaging-test'
+      - '[ "$PPX_LET" != yes ] || make ppx_let-test-deps ppx_let-test'
+
+    before_cache:
+      - opam clean
+
+    cache:
+      directories:
+        - $HOME/.opam
+        - ./_opam
+        - ./_cache
+
+  - &esy
+    install: npm install -g esy
+    script:
+      - esy install
+      - esy build
+      - 'esy sh -c "cat #{self.target_dir}/_build/default/src/unix/unix_c_library_flags.sexp"  | grep lev'
+
+    cache:
+      directories:
+        - $HOME/.esy
+
 matrix:
   include:
-    - os: osx
+    - <<: *opam
+      os: osx
       env: COMPILER=4.09.0 LIBEV=no
-    - os: linux
+    - <<: *opam
+      os: linux
       env: COMPILER=4.09.0 LIBEV=no PPX_LET=yes COVERAGE=yes
-    - os: linux
+    - <<: *opam
+      os: linux
       env: COMPILER=4.08.1+musl+flambda MUSL=yes
-    - os: linux
+    - <<: *opam
+      os: linux
       env: COMPILER=4.07.1 DOCS=yes
-    - os: linux
+    - <<: *opam
+      os: linux
       env: COMPILER=4.06.1 LWT_STRESS_TEST=true
-    - os: linux
+    - <<: *opam
+      os: linux
       env: COMPILER=4.05.0+bytecode-only
-    - os: linux
+    - <<: *opam
+      os: linux
       env: COMPILER=4.04.2
-    - os: linux
+    - <<: *opam
+      os: linux
       env: COMPILER=4.03.0
-    - os: linux
+    - <<: *opam
+      os: linux
       env: COMPILER=4.02.3 PACKAGING=yes
+    - <<: *esy
+      os: linux
+      env: ESY=yes
 
   allow_failures:
```

Comment:
Could you add the esy row to `allow_failures`? Rather than treating these as allowed to fail, I use `allow_failures` for rows that are less likely to fail, given the main rows succeed. Together with `fast_finish`, this gives a reasonably accurate answer from CI sooner than if CI had to wait for all the rows. Rows in `allow_failures` still trigger emails, and I check them manually from time to time and before each release, so we won't miss any regression.